### PR TITLE
Fixed #24915 - Added stricter session key validation

### PR DESCRIPTION
**Pull request by sp1ky**
Changed _session_key attribute to a property and implemented basic
validation in the setter. The session key must be 'truthy' and
at least 8 characters long. Otherwise, the value is set to None.

https://code.djangoproject.com/ticket/24915

### CODE REVIEW

File: tests/sessions_tests/tests.py [hunk 1]
```diff
@@ -198,6 +198,20 @@ def test_invalid_key(self):
             # session key; make sure that entry is manually deleted
             session.delete('1')
 
+    def test_session_key_validation(self):
```
Comment (carljm) [inline tests/sessions_tests/tests.py]: This should be three tests, not one.

**Comment by carljm** [comment]
This looks good to me. At first, the double layer of properties seemed like overkill, but on second thought it seems OK to be cautious by maintaining the `session_key` property as read-only, and avoiding changing the initialization routine, to avoid any issues with third-party session backends.

**Comment by sp1ky** [comment]
@carljm I have split the validation test into three tests as suggested. 

The doubly layered properties look a bit weird all right but I figured it was the ideal way to implement the additional functionality without changing the internal API (ostensibly for the reasons you mentioned, realistically because I don't have a good feel for the code base and took the cautious approach :-P).

**Comment by carljm** [comment]
Looks good to me! Can you squash the second commit into the first (so there's just one commit with the proper commit message) and force push? Then this looks ready for checkin.

**Comment by sp1ky** [comment]
@carljm Done

File: django/contrib/sessions/backends/base.py [hunk 2]
```diff
@@ -161,10 +161,25 @@ def _get_or_create_session_key(self):
             self._session_key = self._get_new_session_key()
         return self._session_key
 
+    def _validate_session_key(self, key):
+        """Key must be truthy and at least 8 characters long"""
```
Comment (timgraham) [inline django/contrib/sessions/backends/base.py]: Please include periods in docstrings.

File: tests/sessions_tests/tests.py [hunk 3]
```diff
@@ -198,6 +198,22 @@ def test_invalid_key(self):
             # session key; make sure that entry is manually deleted
             session.delete('1')
 
+    def test_session_key_empty_string_invalid(self):
+        """Falsey values (Such as an empty string) are rejected"""
+        self.session._session_key = ''
+        self.assertNotEqual(self.session.session_key, '')
```
Comment (timgraham) [inline tests/sessions_tests/tests.py]: not sure I see the need to assertNotEqual when you assert the actual value in the next assertion?

File: tests/sessions_tests/tests.py [hunk 4]
```diff
@@ -198,6 +198,22 @@ def test_invalid_key(self):
...
+        self.assertEqual(self.session.session_key, None)
```
Comment (timgraham) [inline tests/sessions_tests/tests.py]: `assertIsNone`

File: django/contrib/sessions/backends/base.py (hunk 2 above)
Comment (timgraham) [inline django/contrib/sessions/backends/base.py]: Could you explain why the limit is 8? (or mention it's somewhat arbitrary)

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: django/contrib/sessions/backends/base.py (hunk 2 above)
Comment (sp1ky) [inline django/contrib/sessions/backends/base.py]: Added periods to docstrings.
The 8 character limitation is indeed an arbitrary lower bound to enforce a minimum level of collision security. I went with 8 as it was suggested in the trac ticket.

File: tests/sessions_tests/tests.py (hunk 3 above)
Comment (sp1ky) [inline tests/sessions_tests/tests.py]: Good point, I guess I was being unnecessarily explicit. Removed

File: tests/sessions_tests/tests.py (hunk 4 above)
Comment (sp1ky) [inline tests/sessions_tests/tests.py]: Done

**Comment by sp1ky** [comment]
New commit in response to @timgraham's comments. Held off on squashing to allow further development if necessary. Let me know when to re-squash.

**Comment by timgraham** [comment]
Looks good. I added a sentence in the release notes and merged in f4416b1a8b92e492707a6261b7a1132f8550457f, thanks!

**Pull request closed without merging**